server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -377,4 +408,57 @@ public TaskResult result(DiscoveryNode node, ActionResponse response) throws IOE
             throw new IllegalStateException("response has to implement ToXContent to be able to store the results");
         }
     }
+
+    /**
+     * Registers a task resource tracking completion listener on this task.
+     */
+    public void addResourceTrackingCompletionListener(NotifyOnceListener<Task> listener) {
+        resourceTrackingCompletionListeners.add(listener);
+    }
```

Comment:
We shouldn't addResourceTrackingCompletionListener if the count is zero, else its possible to that newly added listener is never called.

Reply:
Fair point, updated.

Reply:
Though there is still a rare possibility of race-condition:
* Task resource tracking is completed (num threads = 0), and existing completion listeners are invoked.
* Delayed thread execution starts for the task (num threads = 1)
* New completion listener added at this point may succeed.
* Delayed thread execution stops for the task (num threads = 0)
* New completion listener is invoked.

To solve this, we may have to bring back the `isResourceTrackingCompleted` atomic boolean into the Task. It's not a concern at the moment as listeners are added during Task creation, not in the middle of execution.

Reply:
On a different note, this may not even be a problem because a (delayed) thread is still a part of the task, and the newly added listener would just receive the more recent/accurate usage stats.

Reply:
Yeah we haven't also strictly synchronised adding listeners and invoking them, so I am fine with this limitation as long as it doesn't overcomplicate the use case